### ISSUE TITLE
CCSprite setSpriteFrame crash when  spriteFrame==nullptr

### ISSUE THREAD

**Issue by InfResearch** [cocos2d/cocos2d-x#14647]
void Sprite::setSpriteFrame(SpriteFrame *spriteFrame)
{
    // retain the sprite frame
    // do not removed by SpriteFrameCache::removeUnusedSpriteFrames
    if (_spriteFrame != spriteFrame)
    {
        CC_SAFE_RELEASE(_spriteFrame);
        _spriteFrame = spriteFrame;
        spriteFrame->retain();
    }
    _unflippedOffsetPositionFromCenter = spriteFrame->getOffset();

```
Texture2D *texture = spriteFrame->getTexture();
// update texture before updating texture rect
if (texture != _texture)
{
    setTexture(texture);
}

// update rect
_rectRotated = spriteFrame->isRotated();
setTextureRect(spriteFrame->getRect(), _rectRotated, spriteFrame->getOriginalSize());

if(spriteFrame->hasPolygonInfo())
{
    _polyInfo = spriteFrame->getPolygonInfo();
}
```

}

If the user forgets to check whether nullptr==spriteFrame
Then the program will crash.

**Comment by stevetranby** [comment]
These type of checks are always a little tricky since you don't want to slow down performance-critical code paths. Curious of other's opinions. It's annoying to have to check outside the method, but often you only need to check during development or if there are resources not included as part of the app package.

**Comment by InfResearch** [comment]
Eh，it hard to debug when i use Lua

**Comment by ricardoquesada** [comment]
yes, cocos2d-x should not crash, even if you are passing invalid parameters to the public API.

but right now it crashes (assert) on purpose. we should change that policy and make it fail safely.
